zkevm-circuits/src/evm_circuit/execution/error_oog_log.rs

```diff
@@ -0,0 +1,366 @@
+                OpcodeId::LOG2.expr(),
+                OpcodeId::LOG3.expr(),
+                OpcodeId::LOG4.expr(),
+            ],
+        );
+        let mstart = cb.query_cell_phase2();
+        let msize = cb.query_word_rlc();
+        let rw_counter_end_of_reversion = cb.query_cell();
+
+        // Pop mstart_address, msize from stack
+        cb.stack_pop(mstart.expr());
+        cb.stack_pop(msize.expr());
+
+        // Note: no need to check not in static call, since write protection error will
+        // handle it.
+        let topic_count = opcode.expr() - OpcodeId::LOG0.as_u8().expr();
+
+        // check memory
+        let memory_address = MemoryAddressGadget::construct(cb, mstart, msize);
+
+        // Calculate the next memory size and the gas cost for this memory
+        // access
+        let memory_expansion = MemoryExpansionGadget::construct(cb, [memory_address.address()]);
+
+        let gas_cost = GasCost::LOG.as_u64().expr()
+            + GasCost::LOG.as_u64().expr() * topic_count
+            + 8.expr() * memory_address.length()
+            + memory_expansion.gas_cost();
+
+        // Check if the amount of gas available is less than the amount of gas
+        // required
+        let insufficient_gas = LtGadget::construct(cb, cb.curr.state.gas_left.expr(), gas_cost);
+        cb.require_equal(
+            "log* gas left is less than gas required ",
+            insufficient_gas.expr(),
+            1.expr(),
+        );
+
+        // current call must be failed.
+        cb.call_context_lookup(false.expr(), None, CallContextFieldTag::IsSuccess, 0.expr());
+
+        cb.call_context_lookup(
+            false.expr(),
+            None,
+            CallContextFieldTag::RwCounterEndOfReversion,
+            rw_counter_end_of_reversion.expr(),
+        );
+
+        // Go to EndTx only when is_root
+        let is_to_end_tx = cb.next.execution_state_selector([ExecutionState::EndTx]);
+        cb.require_equal(
+            "Go to EndTx only when is_root",
+            cb.curr.state.is_root.expr(),
+            is_to_end_tx,
+        );
+
+        // When it's a root call
+        cb.condition(cb.curr.state.is_root.expr(), |cb| {
+            // Do step state transition
+            cb.require_step_state_transition(StepStateTransition {
+                call_id: Same,
+                rw_counter: Delta(4.expr() + cb.curr.state.reversible_write_counter.expr()),
+                ..StepStateTransition::any()
+            });
+        });
+
+        // When it's an internal call, need to restore caller's state as finishing this
+        // call. Restore caller state to next StepState
+        let restore_context = cb.condition(1.expr() - cb.curr.state.is_root.expr(), |cb| {
+            RestoreContextGadget::construct(
+                cb,
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+                0.expr(),
+            )
```

Comment:
We could maybe create an associated fn for `RestoreContextGadget` like: `RestoreContextGadget::from_revert_ctx` which directly places all the 0's and has less "magic numbers" arround.

Reply:
good idea , better to do it with common error gadget refactoring in https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1059